phoenix-core/src/main/java/org/apache/phoenix/expression/util/regex/AbstractBaseMatcher.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.util.regex;
+
+public abstract class AbstractBaseMatcher {
```

Comment:
Actually, simplify this by not passing in SortOrder. Instead use the dataType.coerceBytes() call I mentioned already (in your Utils class) to invert before calling matcher. That way the matcher doesn't need to know anything about SortOrder. So just have this:

    boolean matcher(ImmutableBytesWritable ptr);